File: python/day12.py
from aoc import *
from collections import deque


def parse_line(line):
    return [ord(c) - ord('a') for c in line]


def find_start_end(grid):
    start = next(k for k, v in grid.items() if v == -14)
    grid[start] = 0
    end = next(k for k, v in grid.items() if v == -28)
    grid[end] = 25
    return start, end


def travel(grid, start, end, part):
    seen = set()
    queue = deque([(0, end)])
    while queue:
        steps, curr = queue.popleft()
        if grid[curr] == 0 and (part == 2 or curr == start):
            return steps
        for nb in neighbours(*curr):
            if nb not in seen and grid.get(nb, -inf) >= grid[curr] - 1:
                seen.add(nb)
                queue.append((steps+1, nb))
    return steps


def solve(file=12):
    grid = list2grid(map(parse_line, read_input(file)))
    start, end = find_start_end(grid)
    return (travel(grid, start, end, 1),
            travel(grid, start, end, 2))


print(solve())










def test():
    assert solve("12_test") == (31, 29)
    assert solve() == (352, 345)

File: clojure/day12.clj
(ns day12
  (:require aoc
            [clojure.data.int-map :refer [int-map]]))


(def ^:const N 100)


(defn neighbours ^longs [^long pt]
  (for [^long delta [-1 1 (- N) N]]
    (+ delta pt)))

(defn find-val [grid height]
  (some #(when (= height (val %)) (key %)) grid))

(defn travel [grid part]
  (let [start (find-val grid -14)
        end   (find-val grid -28)
        grid  (-> grid
                  (assoc start 0)
                  (assoc end 25))]
    (loop [seen  #{end}
           queue [[0 end]]]
      (let [[steps curr] (peek queue)
            h (grid curr)]
        (if (and (zero? h)
                 (or (= part 2) (= curr start)))
          steps
          (let [nbs (filter #(and (not (seen %))
                                  (>= (grid % -99) (dec h)))
                            (neighbours curr))
                nexts (mapv #(vector (inc steps) %) nbs)]
            (recur (into seen nbs)
                   (into nexts (pop queue)))))))))

(defn make-grid [lines]
  (into (int-map)
        (for [[y line] (map-indexed vector lines)
              [x char] (map-indexed vector line)]
          [(+ x (* N y))
           (- (int char) (int \a))])))

(defn solve
  ([] (solve (aoc/read-file 12)))
  ([input]
   (let [grid (make-grid (aoc/parse-input input :chars))
         p1 (future (travel grid 1))
         p2 (future (travel grid 2))]
     [@p1 @p2])))


(solve)
